File: msr/reference/typescript/run-length-encoding.ts
function rle_encode(s: string): string {
  let out = "";
  let i = 0;
  while (i < s.length) {
    let j = i;
    while (j < s.length && s[j] === s[i]) j++;
    out += String(j - i) + s[i];
    i = j;
  }
  return out;
}

function rle_decode(s: string): string {
  let out = "";
  for (let i = 0; i + 1 < s.length; i += 2) out += s[i + 1].repeat(Number(s[i]));
  return out;
}

File: msr/reference/go/run-length-encoding.go
package main

import (
	"strconv"
	"strings"
)

func rle_encode(s string) string {
	var b strings.Builder
	for i := 0; i < len(s); {
		j := i
		for j < len(s) && s[j] == s[i] {
			j++
		}
		b.WriteString(strconv.Itoa(j - i))
		b.WriteByte(s[i])
		i = j
	}
	return b.String()
}

func rle_decode(s string) string {
	var b strings.Builder
	for i := 0; i+1 < len(s); i += 2 {
		n := int(s[i] - '0')
		b.WriteString(strings.Repeat(string(s[i+1]), n))
	}
	return b.String()
}
